Add YTT symlink options and shell quote ytt command arguments

loricvdt opened the pull request:
Adds two environment variables to allow YTT templating using symlinks, mapped to the ytt CLI options:
```bash
YTT_ALLOW_SYMLINK_DESTINATIONS=/paths/to/allow # Mapped to --allow-symlink-destination
YTT_DANGEROUS_ALLOW_ALL_SYMLINK_DESTINATIONS=true # Mapped to --dangerous-allow-all-symlink-destinations
```
Following this: https://carvel.dev/ytt/docs/latest/faq/#how-can-i-use-files-that-are-symlinks

And adds shell quoting to ytt CLI args in the process.

ansibleguy76 commented:
First : thank you for the pull request !

is this pull request against main of devel ?  As I use devel branch for development
I also see you use the env vars directly in the code.  
I'd rather like to see them processed in the config code.
but I also see you loop some dynamic env vars ?  (YTT_LIB_DATA_) ?  Can you explain ?

ansibleguy76 reviewed (changes requested):
Can you write proper help for all variables ?  Also the dynamic ones ?

ansibleguy76 commented:
example of help
```

  - name: YTT_ALLOW_SYMLINK_DESTINATIONS
    type: string
    allowed: valid paths
    short: Ytt allowed symlink destinations
    default: ""
    description: |
      Ytt does not allow symlink destinations by default.  If you want to allow symlink destinations, you can set this variable to the path where the symlinks are allowed.
    version: 5.0.8
  - name: YTT_DANGEROUS_ALLOW_ALL_SYMLINK_DESTINATIONS
    type: number
    choices:
      - name: 0
        description: Disables all symlink destinations
      - name: 1
        description: Enables all symlink destinations
    default: 0
    short: Ytt dangerous allow all symlink destinations
    description: |
      Ytt does not allow symlink destinations by default.  If you want to allow all symlink destinations, you can set this variable to 1.
    version: 5.0.8
```

ansibleguy76 commented:
```
Object.entries(process.env).filter( ([key, value]) => key.startsWith('YTT_LIB_DATA_')).forEach( ([key, value]) => {
  const libName = key.replace('YTT_LIB_DATA_', '').toLowerCase();
  app_config.yttLibData.push({libName, value});
});
```
I will preprocess dynamic vars so that the code has an array of objects

ansibleguy76 commented:
I believe this is now correctly implemented in beta 5.0.8 (to be released soon)

ansibleguy76 reviewed (commented):
```
  - name: YTT_ALLOW_SYMLINK_DESTINATIONS
    type: string
    allowed: valid paths
    short: Ytt allowed symlink destinations
    default: ""
    description: |
      Ytt does not allow symlink destinations by default.  If you want to allow symlink destinations, you can set this variable to the path where the symlinks are allowed.
    version: 5.0.8
  - name: YTT_DANGEROUS_ALLOW_ALL_SYMLINK_DESTINATIONS
    type: number
    choices:
      - name: 0
        description: Disables all symlink destinations
      - name: 1
        description: Enables all symlink destinations
    default: 0
    short: Ytt dangerous allow all symlink destinations
    description: |
      Ytt does not allow symlink destinations by default.  If you want to allow all symlink destinations, you can set this variable to 1.
    version: 5.0.8
  - name: YTT_VARS_PREFIX
    type: string
    allowed: a valid prefix
    short: Ytt vars prefix
    default: ""
    description: |
      Ytt does not allow vars by default.  If you want to allow vars, you can set this variable to the prefix that is allowed.
    version: 5.0.8
  - name: YTT_LIB_DATA_{dynamic}
    type: string
    allowed: a valid lib data value
    short: Ytt lib data
    default: ""
    description: |
      This is not a single variable but rather an infinite amount of variables.  You can set as many as you want.  The dynamic part is the name of the lib data.  For example `YTT_LIB_DATA_MYLIB` will be used as lib data in ytt.
    version: 5.0.8
```

ansibleguy76 commented:
I will await your feedback if this help is correct

I have moved the processing of the variables to app.config.js
and adopted forms.model & help

loricvdt commented:
Sounds good, here's a bit of rewording and clarification for the help (feel free to tweak it if you wish):
```
  - name: YTT_ALLOW_SYMLINK_DESTINATIONS
    type: string
    allowed: valid paths
    short: Ytt allowed symlink destinations
    default: ""
    description: |
      Ytt disables templating with symlinks by default.  If you want to allow symlink use on specific directories, this variable can be set to the path(s) where the symlinks are allowed (multiple paths separated by Node's `path.delimiter`). Note: enabling this may come with some risks, see ytt FAQ for more info.
    version: 5.0.8
  - name: YTT_DANGEROUS_ALLOW_ALL_SYMLINK_DESTINATIONS
    type: number
    choices:
      - name: 0
        description: Disables 'allow all symlink destinations'
      - name: 1
        description: Enables 'allow all symlink destinations'
    default: 0
    short: Ytt dangerous allow all symlink destinations
    description: |
      Ytt disables templating with symlinks by default.  If you want to allow all symlink use EVERYWHERE, you can set this variable to 1. Note: enabling this comes with some risks (hence the 'DANGEROUS' part), see ytt FAQ for more info.
    version: 5.0.8
  - name: YTT_VARS_PREFIX
    type: string
    allowed: a valid prefix
    short: Ytt vars prefix
    default: ""
    description: |
      Environment variable prefix for ytt to get data values from. For example, when set to `MY_PREFIX`, `MY_PREFIX_my_var=value` results to the ytt data value `my_var=value`.
    version: 5.0.8
  - name: YTT_LIB_DATA_{dynamic}
    type: string
    allowed: a valid lib data value
    short: Ytt lib data
    default: ""
    description: |
      This is not a single variable but rather an infinite amount of variables.  You can set as many as you want.  The dynamic part is the name of the lib data.  For example `YTT_LIB_DATA_MYLIB=values.yml` results in the contents of `value.yml` to be used by ytt in the `mylib` library.
    version: 5.0.8
```

Not sure how you managed the defaults, but just a heads up that the code I wrote does not handle empty string ("") as a default so some stuff may break if you kept the same logic (such as the `.split(regex)` part)

I made the PR onto develop, I was not sure which to use so I looked at the previous PRs which were also onto develop.
I added the vars there because the previous YTT code was there so I wanted to stay consistent, but I agree that it is better with the rest of the config.

I will test your beta once you publish it.
Thanks for the quick reaction

ansibleguy76 commented:
added to 5.0.8, closing this... please test the new beta if you please.

ansibleguy76 commented:
Hi,

HAve you been able to test the new beta ?

Op do 2 jan 2025 om 15:37 schreef Loric Vandentempel <
***@***.***>:

> Adds two environment variables to allow YTT templating using symlinks,
> mapped to the ytt CLI options:
>
> YTT_ALLOW_SYMLINK_DESTINATIONS=/paths/to/allow # Mapped to --allow-symlink-destination
> YTT_DANGEROUS_ALLOW_ALL_SYMLINK_DESTINATIONS=true # Mapped to --dangerous-allow-all-symlink-destinations
>
> Following this:
> https://carvel.dev/ytt/docs/latest/faq/#how-can-i-use-files-that-are-symlinks
>
> And adds shell quoting to ytt CLI args in the process.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/ansibleguy76/ansibleforms/pull/246
> Commit Summary
>
>    - 5dc4d6d
>    <https://github.com/ansibleguy76/ansibleforms/pull/246/commits/5dc4d6d1b6308ae086861509e8738cfce1afe6fc>
>    Add YTT symlink options and shell quote ytt command arguments
>
> File Changes
>
> (2 files <https://github.com/ansibleguy76/ansibleforms/pull/246/files>)
>
>    - *M* docs/_data/help.yaml
>    <https://github.com/ansibleguy76/ansibleforms/pull/246/files#diff-e16e2b31b3d685cd298aa92dba45485d0677be8a6dde14cfcbd9c062bcffb6a3>
>    (14)
>    - *M* server/src/models/form.model.js
>    <https://github.com/ansibleguy76/ansibleforms/pull/246/files#diff-4d917f67e8db63895c81217968a6d58e2e09db2595081ad979aa227b868cecbb>
>    (38)
>
> Patch Links:
>
>    - https://github.com/ansibleguy76/ansibleforms/pull/246.patch
>    - https://github.com/ansibleguy76/ansibleforms/pull/246.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ansibleguy76/ansibleforms/pull/246>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVM6SKS4CLDLEOLODRHK7RT2IVFKZAVCNFSM6AAAAABUQCG73KVHI2DSMVQWIX3LMV43ASLTON2WKOZSG43DMMJTHE2TGNY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
